File: magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/external/registry/RegisterWebhook.scala
package au.csiro.data61.magda.indexer.external.registry

import scala.concurrent.Future
import akka.actor.ActorSystem
import akka.stream.Materializer
import com.typesafe.config.Config
import scala.concurrent.ExecutionContext
import au.csiro.data61.magda.model.Registry.{ WebHook, EventType, WebHookConfig }
import au.csiro.data61.magda.client.RegistryExternalInterface
import au.csiro.data61.magda.model.Registry.RegistryConstants
import akka.event.LoggingAdapter
import akka.event.Logging

object RegisterWebhook {
  sealed trait InitResult
  case object ShouldCrawl extends InitResult
  case object ShouldNotCrawl extends InitResult

  def initWebhook(interface: RegistryExternalInterface)(
    implicit config: Config,
    system: ActorSystem,
    executor: ExecutionContext,
    materializer: Materializer): Future[InitResult] = {

    val logger = Logging(system, getClass)

    logger.info("Looking up existing webhook with id {}", config.getString("registry.webhookId"))
    interface.getWebhook(config.getString("registry.webhookId")).flatMap {
      case Some(existingHook) =>
        logger.info("Hook already exists, updating...")
        registerIndexerWebhook(interface, RegistryConstants.aspects, RegistryConstants.optionalAspects, true)
          .map { _ =>
            logger.info("Updated, attempting to resume...")
            interface.resumeWebhook(config.getString("registry.webhookId"))
          } map { _ =>
            logger.info("Successfully resumed webhook")

            ShouldNotCrawl
          }
      case None =>
        logger.info("No hook exists, registering a new one")
        registerIndexerWebhook(interface, RegistryConstants.aspects, RegistryConstants.optionalAspects)
          .map { _ =>
            logger.info("Successfully registered new webhook")

            ShouldCrawl
          }
    }
  }

  private def registerIndexerWebhook(
    interface: RegistryExternalInterface,
    aspects: List[String],
    optionalAspects: List[String],
    isUpdate: Boolean = false)(implicit config: Config, system: ActorSystem, executor: ExecutionContext, materializer: Materializer): Future[Unit] = {
    val webhook = WebHook(
      id = Some(config.getString("registry.webhookId")),
      name = "Indexer",
      eventTypes = Set(
        EventType.CreateRecord,
        EventType.PatchRecordAspect,
        EventType.DeleteRecord,
        EventType.CreateRecordAspect,
        EventType.PatchRecord,
        EventType.DeleteRecordAspect),
      url = config.getString("registry.webhookUrl"),
      config = WebHookConfig(
        aspects = Some(aspects),
        optionalAspects = Some(optionalAspects),
        includeEvents = Some(true),
        includeRecords = Some(true),
        dereference = Some(true)),
      userId = Some(0), // TODO: Will have to change this when it becomes important
      isWaitingForResponse = None,
      active = true)


    val doRegister = if (isUpdate) {
      interface.putWebhook(webhook)
    } else {
      interface.createWebhook(webhook)
    }

    doRegister.map { _ =>
      system.log.info("Successfully added webhook")

      Unit
    }.recover {
      case e: Throwable =>
        system.log.error(e, "Failed to add webhook")
        throw e
    }.map(_ => Unit)
  }
}

File: magda-sleuther-framework/src/registerWebhook.ts
import {
    WebHook,
    WebHookConfig
} from "@magda/typescript-common/dist/generated/registry/api";
import Registry from "@magda/typescript-common/dist/registry/AuthorizedRegistryClient";

import SleutherOptions from "./SleutherOptions";
import getWebhookUrl from "./getWebhookUrl";

export type NeedsCrawl = Promise<boolean>;

export default async function registerNewWebhook(
    options: SleutherOptions,
    registry: Registry
) {
    console.info("Registering webhook");

    const webHookConfig: WebHookConfig = {
        aspects: options.aspects,
        optionalAspects: options.optionalAspects,
        includeEvents: false,
        includeRecords: true,
        includeAspectDefinitions: false,
        dereference: true
    };

    const newWebHook: WebHook = {
        id: options.id,
        userId: 0, // TODO: When this matters
        name: options.id,
        active: true,
        url: getWebhookUrl(options),
        eventTypes: [
            "CreateRecord",
            "CreateAspectDefinition",
            "CreateRecordAspect",
            "PatchRecord",
            "PatchAspectDefinition",
            "PatchRecordAspect"
        ],
        config: webHookConfig,
        lastEvent: null,
        isWaitingForResponse: false,
        lastRetryTime: null,
        retryCount: 0
    };

    return registry.postHook(newWebHook);
}
